common/chaton.hpp

```diff
@@ -0,0 +1,758 @@
+            lens.push_back(part.length());
+        }
+        return lens;
+    }
+
+    std::string name() {
+        return typeid(*this).name();
+    }
+
+    void dump() {
+        std::string me = name() + ":" + __func__;
+        LOGXLN("INFO:%s:NumTypes:%zu", me.c_str(), types.length());
+        LOGXLN("INFO:%s:NumParts:%zu", me.c_str(), parts.size());
+        LOGXLN("INFO:%s:StrLength:%zu", me.c_str(), str().length());
+        if (parts.size() != types.length()) {
+            LOG_TEELN("DBUG:%s:Mismatch between parts and types", me.c_str());
+        }
+        int i = 0;
+        for(auto part: parts) {
+            LOGXLN("INFO:%s:%c:%s", me.c_str(), types[i], part.c_str());
+            i += 1;
+        }
+    }
+
+};
+
+inline bool chaton_meta_load(std::string &fname) {
+    if (conMeta != nullptr) {
+        LOGXLN("WARN:%s:ChatOn Meta: overwriting???", __func__);
+    }
+    std::ifstream f(fname);
+    conMeta = json::parse(f);
+    return true;
+}
+
+inline bool chaton_tmpl_exists(const std::string &tmpl) {
+    if (conMeta == nullptr) {
+        LOG_TEELN("ERRR:%s:ChatOnMeta: Not loaded yet...", __func__);
+        return false;
+    }
+    try {
+        auto tmplData = conMeta[tmpl];
+        return true;
+    } catch (json::exception &err) {
+        LOG_TEELN("WARN:%s:ChatOnMeta: tmpl[%s] not found...", __func__, tmpl.c_str());
+        return false;
+    }
+}
+
+inline std::string chaton_tmpl_role_kv(const std::string &tmpl, const std::string &role, const std::vector<std::string> &keys) {
+    std::string got = "";
+    std::string sKeys = "";
+    for(auto key: keys) {
+        try {
+            got += conMeta[tmpl][role][key];
+        } catch (json::exception &err) {
+        }
+        sKeys += "+";
+        sKeys += key;
+    }
+    LOGLN("DBUG:%s:%s:%s:%s:%s", __func__, tmpl.c_str(), role.c_str(), sKeys.c_str(), got.c_str());
+    return got;
+}
+
+inline std::string chaton_tmpl_kv(const std::string &tmpl, const std::string &key) {
+    std::string got = conMeta[tmpl][key];
+    LOGLN("DBUG:%s:%s:%s:%s", __func__, tmpl.c_str(), key.c_str(), got.c_str());
+    return got;
+}
+
+inline bool chaton_tmpl_kv_bool(const std::string &tmpl, const std::string &key) {
+    bool got = conMeta[tmpl][key];
+    LOGLN("DBUG:%s:%s:%s:%d", __func__, tmpl.c_str(), key.c_str(), got);
+    return got;
+}
+
+
+// Given the template standard, role and a message, this returns
+// a tagged message, types string and lens vector wrt the parts that make up the returned string
+//
+// * a string containing the tagged message
+//   * role-(begin+prefix) + msg + role-(suffix+end)
+// * a string where the chars contain info about
+//   type of sub-strings/parts that make up the tagged message.
+// * a vector of ints, which give the length of each part in the tagged message.
+inline bool chaton_tmpl_apply_single_ex(
+        const std::string &tmpl,
+        const std::string &role,
+        const std::string &content,
```

Comment:
We put spaces front and back:

```suggestion
        const std::string & tmpl,
        const std::string & role,
        const std::string & content,
```

Reply:
understood